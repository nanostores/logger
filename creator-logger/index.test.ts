import { delay } from 'nanodelay'
import type { MapCreator, MapStore } from 'nanostores'
import { map, onMount, STORE_UNMOUNT_DELAY } from 'nanostores'
import { afterEach, beforeAll, expect, it, vi } from 'vitest'

import { action } from '../action/index.js'
import { format } from '../test/index.js'
import { creatorLogger } from './index.js'

function createMapCreator(
  init?: (store: MapStore, id: string) => void
): MapCreator {
  let Template: any = (id: string) => {
    if (!Template.cache[id]) {
      Template.cache[id] = Template.build(id)
    }
    return Template.cache[id]
  }

  Template.build = (id: string) => {
    let store = map({ id, value: null })
    onMount(store, () => {
      if (init) init(store, id)
      return () => {
        delete Template.cache[id]
      }
    })
    return store
  }

  Template.cache = {}

  return Template
}

let out = ''
let groups = 0

beforeAll(() => {
  vi.spyOn(console, 'groupCollapsed').mockImplementation((...args: any[]) => {
    console.log(...args)
    groups += 1
  })
  vi.spyOn(console, 'log').mockImplementation((...args: any[]) => {
    if (out === '') out += '\n'
    out += format(groups, [], ...args) + '\n'
  })
  vi.spyOn(console, 'groupEnd').mockImplementation(() => {
    groups -= 1
  })
})

afterEach(() => {
  out = ''
  groups = 0
})

it('prints logs from store builder', async () => {
  let creator = createMapCreator($store => {
    $store.setKey('value', 0)
    let increaseValue = action($store, 'Increase Value', store => {
      store.setKey('value', Number(store.get().value) + 1)
    })
    increaseValue()
    increaseValue()
    increaseValue()
  })
  let destroy = creatorLogger({ 'Map Creator': creator })

  let $first = creator('1')
  let $second = creator('2')
  let $third = creator('3')
  let unbindFirst = $first.listen(() => {})
  let unbindSecond = $second.listen(() => {})
  let unbindThird = $third.listen(() => {})
  unbindFirst()
  unbindSecond()
  unbindThird()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})

it('returns unbind function', () => {
  let creator = createMapCreator()
  let destroy = creatorLogger({ 'Map Creator': creator })
  destroy()

  let $store = creator('1')
  let unbindStore = $store.listen(() => {})
  unbindStore()

  expect(out).toBe('')
})
