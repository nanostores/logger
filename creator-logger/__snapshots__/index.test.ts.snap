// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prints logs from store builder 1`] = `
"
Nano Stores build: Map Creator:1 store was built by Map Creator creator
Nano Stores build: Map Creator:2 store was built by Map Creator creator
Nano Stores build: Map Creator:3 store was built by Map Creator creator
Nano Stores mount: Map Creator:1 store was mounted
Nano Stores change: Map Creator:1 store was changed in the value key
  value: null → 0 
  new: {"id":"1","value":0}
  old: {"id":"1","value":null}
Nano Stores action: Map Creator:1 store was changed by action Increase Value
  change: Map Creator:1 store was changed in the value key
    value: 0 → 1 
    new: {"id":"1","value":1}
    old: {"id":"1","value":0}
Nano Stores action: Map Creator:1 store was changed by action Increase Value
  change: Map Creator:1 store was changed in the value key
    value: 1 → 2 
    new: {"id":"1","value":2}
    old: {"id":"1","value":1}
Nano Stores action: Map Creator:1 store was changed by action Increase Value
  change: Map Creator:1 store was changed in the value key
    value: 2 → 3 
    new: {"id":"1","value":3}
    old: {"id":"1","value":2}
Nano Stores mount: Map Creator:2 store was mounted
Nano Stores change: Map Creator:2 store was changed in the value key
  value: null → 0 
  new: {"id":"2","value":0}
  old: {"id":"2","value":null}
Nano Stores action: Map Creator:2 store was changed by action Increase Value
  change: Map Creator:2 store was changed in the value key
    value: 0 → 1 
    new: {"id":"2","value":1}
    old: {"id":"2","value":0}
Nano Stores action: Map Creator:2 store was changed by action Increase Value
  change: Map Creator:2 store was changed in the value key
    value: 1 → 2 
    new: {"id":"2","value":2}
    old: {"id":"2","value":1}
Nano Stores action: Map Creator:2 store was changed by action Increase Value
  change: Map Creator:2 store was changed in the value key
    value: 2 → 3 
    new: {"id":"2","value":3}
    old: {"id":"2","value":2}
Nano Stores mount: Map Creator:3 store was mounted
Nano Stores change: Map Creator:3 store was changed in the value key
  value: null → 0 
  new: {"id":"3","value":0}
  old: {"id":"3","value":null}
Nano Stores action: Map Creator:3 store was changed by action Increase Value
  change: Map Creator:3 store was changed in the value key
    value: 0 → 1 
    new: {"id":"3","value":1}
    old: {"id":"3","value":0}
Nano Stores action: Map Creator:3 store was changed by action Increase Value
  change: Map Creator:3 store was changed in the value key
    value: 1 → 2 
    new: {"id":"3","value":2}
    old: {"id":"3","value":1}
Nano Stores action: Map Creator:3 store was changed by action Increase Value
  change: Map Creator:3 store was changed in the value key
    value: 2 → 3 
    new: {"id":"3","value":3}
    old: {"id":"3","value":2}
Nano Stores unmount: Map Creator:1 store was unmounted
Nano Stores unmount: Map Creator:2 store was unmounted
Nano Stores unmount: Map Creator:3 store was unmounted
"
`;
